### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by tinchox5** [zumerlab/zumly#8] [Ideas]
This is a flex roadmap open to discussion. Ideas and help are welcome!!

- [x] Create new events at the start and end of zoom transitions to allow developers to control how new views are rendered. Check DEV branch
- [x] Optimize Zumly code, simplify API. Check DEV branch
- [x] Adapt Web Animations Api (WAAPI) to the Zumly engine. Check DEV branch
- [ ] Allow users to parameterize zoom transition options for each view. For example, delay, endDelay, easing, duration, etc.
- [ ] Allow users to determine the relative location of the new transition. Currently, the new view is positioned at the center of the zoomed container, but it could be defined to be positioned according to the coordinates of the zoomed element.
- [ ] Implement lateral navigation at the same zoom level. **I'm working on it**
- [ ] Create a plugin system architecture.
- [ ] Create plugins to use third party animation libraries. **first implement plugin arq., then abstract transform params into vars to be used by third animations libraries**
- [ ] Create a router plugin
- [ ] Design plugins to render views compatible with popular frameworks like React, Vue, Svelte, Angular, etc.
- [ ] Build Zumly's web-components or wrappers to allow importing Zumly into the most popular JS frameworks.
- [ ] Implement scheduled navigation for presentations or automations.
- [ ] Use Vite for the development environment.
- [ ] Document Zumly features
- [ ] Adapt Zircle UI to the Zumly engine.
- [ ] Adapt Zumly engine to work with html canvas. **first implement plugin arq., then abstract transform params into vars to be used by third animations libraries**
- [ ] Use attributes instead of data-*. Example, zumly-easing="ease-in-out" instead of data-easing="ease-in-out"
- [ ] Better control how to handle new views that have different sizes than previous ones. Currently, there is an undocumented 'cover' option that tries to fit the new view to the width or height of the old one.
- [ ] Allow a full zoom transition from one view to the other, which makes it unnecessary to keep the other views visible.
- [ ] [**Community**] Add gesture navigation for mobile devices. Zumly could use pinch zoom functionality based on gestures for touch screens

**Comment by tinchox5** [comment]
**Thoughts about side-navigation.**
- Side-navigation improves UX
- It keeps elements in same context.
- side-nav will be optional
- **renderView()** needs to be modified. 
a) When an element is zoomed,  a new css class `side-nav` is applied to its zoomable siblings. 
b) When a `side-nav`element is zoomed, a new view is rendered and centered in the zoom container but all previous views remain intact. 
c) The new "side view" is added into the storeViews with a 'side-view'  tag
d) Optionally, all side-views remain visible.
- **zoomIn()**, **zoomOut()**, and **key events listeners** need to be modified.

**Comment by camicase82** [comment]
Add gesture navigation for mobile devices

Zumly could use pinch zoom functionality based on gestures for touch screens

**Reply by tinchox5** [reply]
Awesome. Added to roadmap

**Comment by tinchox5** [comment]
To achieve at least goals "**enable third animations libraries**" & "**use zumly in html canvas**", it is necessary to abstract coordinates and scale transform properties, which is good because zumly becomes flexible

**Comment by mdomedia** [comment]
Would love to see keyboard navigation implemented for accessibility
